src/lib.rs

```diff
@@ -308,6 +309,16 @@ impl Core {
                 test_println!("channel full");
                 return Err(TrySendError::Full(()));
             }
+            // if the current state is marked as data written by the current (or even future)
+            // generation we can spin as probably our tail is behind the actual tail
+            // but if the state is marked as data written by the previous generation and nobody
+            // is reading it then we don't have any available slots
+            let (tail_idx, tail_gen) = self.idx_gen(tail);
+            let (state_ix, state_gen) = self.idx_gen(state);
+            if test_dbg!(state_ix == tail_idx + 1 && state_gen < tail_gen) {
```

Comment:
this is very much not a big deal, as it only impacts the traces output in tests, but i might consider splitting these into two separate `test_dbg!` macro invocations, so that in the case where one is true and the other is false, we see which side of the `&&` was false:
```suggestion
            let (state_ix, state_gen) = self.idx_gen(state);
            if test_dbg!(state_ix == tail_idx + 1) && test_dbg!(state_gen < tail_gen) {
```